homeassistant/components/withings/common.py

```diff
@@ -37,40 +50,82 @@ class ServiceError(HomeAssistantError):
 class ThrottleData:
     """Throttle data."""
 
-    def __init__(self, interval: int, data):
+    def __init__(self, interval: int, data: Any):
         """Constructor."""
         self._time = int(time.time())
         self._interval = interval
         self._data = data
 
     @property
-    def time(self):
+    def time(self) -> int:
         """Get time created."""
         return self._time
 
     @property
-    def interval(self):
+    def interval(self) -> int:
         """Get interval."""
         return self._interval
 
     @property
-    def data(self):
+    def data(self) -> Any:
         """Get data."""
         return self._data
 
-    def is_expired(self):
+    def is_expired(self) -> bool:
         """Is this data expired."""
         return int(time.time()) - self.time > self.interval
 
 
+class ConfigEntryWithingsApi(AbstractWithingsApi):
+    """Withing API that uses HA resources."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: ConfigEntry,
+        implementation: AbstractOAuth2Implementation,
+    ):
+        """Initialize object."""
+        self._hass = hass
+        self._config_entry = config_entry
+        self._implementation = implementation
+        self.session = OAuth2Session(hass, config_entry, implementation)
+
+    def _request(
+        self, path: str, params: Dict[str, Any], method: str = "GET"
+    ) -> Dict[str, Any]:
+        return run_coroutine_threadsafe(
+            self.async_do_request(path, params, method), self._hass.loop
+        ).result()
+
+    async def async_do_request(
+        self, path: str, params: Dict[str, Any], method: str = "GET"
+    ) -> Dict[str, Any]:
+        """Perform an async request."""
+        await self.session.async_ensure_token_valid()
+
+        response = await self._hass.async_add_executor_job(
+            partial(
+                requests.request,
```

Comment:
@balloob This is not in line with one of the most important foundations of Home Assistant. Device specific interface should be in a standalone library.

Reply:
(We might have to move this to an architecture issue)

With the new OAuth2 helper, we have authentication tokens that are owned by Home Assistant. Home Assistant is responsible for keeping them up to date and storing them (in a config entry). Renewing is done by calling the `async_ensure_token_valid()` method on an `OAuth2Session` instance. It will make sure the tokens are up to date, and if an update was necessary, make sure the updated tokens are stored inside the config entry.

However, with the authentication living inside Home Assistant, we need to be able to expose this to the PyPI lib. The way it is done here is as far as I know the right approach:
 - The lib exposes an abstract auth class. This class is responsible for making authenticated requests. It does not know about API, implementation or anything else.
 - We implement this auth class and tie it into the `OAuth2Session`
 - We pass the auth class to the actual API lib of the PyPI package.

The only thing that is inside Home Assistant is how the authentication needs to be attached to the request. I guess the alternative could be that we have the auth class expect a method called `get_tokens()` that returns up to date tokens and the auth lib does the request itself.

Reply:
Why do we need to put the authentication low level part inside Home Assistant? Wouldn't storing the tokens be enough and let the library handle renewal?

Reply:
Renewal can only be done inside the library if the library has access to the client ID and client secret. I am about to open a PR (waiting for after beta is cut) that adds a new service for Home Assistant users where Home Assistant owns the client secret, and thus has to be part of the renew call.

Reply:
Would that disallow passing the client id and secret through to the library?

Reply:
@vangorra discussed it with Martin. He is right and we should move to where we only have to pass in `get_tokens()` and we deal with the rest inside the lib. I will update Somfy to show that as an example. For now this PR can be merged and we can update that in the future.